### PR TITLE
[query] Add `isIsomorphicTo` for virtual types

### PR DESCRIPTION
**Pull request by ehigham**
It's useful to know when types are isomorphic to one-another for cast-rename operations.
We had a method of the same name in `TBaseStruct` that did't really do isomorphism as it included names in field types. 
I've renamed this to `isJoinableWith` as a consequence.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -13,7 +13,7 @@ import is.hail.expr.ir.streams.{EmitStream, StreamProducer, StreamUtils}
 import is.hail.io.{BufferSpec, InputBuffer, OutputBuffer, TypedCodecSpec}
 import is.hail.io.fs.FS
 import is.hail.linalg.{BLAS, LAPACK, LinalgCodeUtils}
-import is.hail.types.{TypeWithRequiredness, VirtualTypeWithReq, tcoerce}
+import is.hail.types.{tcoerce, TypeWithRequiredness, VirtualTypeWithReq}
```
Comment (ehigham) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:16]: I'm not sure why we're still seeing these kinds of change @patrick-schultz - I thought we had run scalafix on the whole codebase?

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:16]: Ah, looks like I forgot to add a scalafix check in the `check-hail` make target. I'll PR. I just tried running scalafix locally and this was the only failure that has slipped in.

Comment (ehigham) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:16]: https://github.com/hail-is/hail/pull/14276

**Review by chrisvittal — COMMENTED** [review]
One question, otherwise LGTM.

**Review by chrisvittal — CHANGES_REQUESTED** [review]
Also resolve your merge conflicts

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStackStruct.scala [hunk 2]
```diff
@@ -161,6 +161,12 @@ final case class SStackStruct(virtualType: TBaseStruct, fieldEmitTypes: IndexedS
       ts.types.zip(fieldEmitTypes).map { case (v, e) => e.copy(st = e.st.castRename(v)) },
     )
   }
+
+  override def isIsomorphicTo(st: SType): Boolean =
+    st match {
+      case s: SStackStruct => virtualType isJoinableWith s.virtualType
```
[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStackStruct.scala:167]: Ah good spot, that's a bug!

[the remark this answers is not in the document]
Comment (ehigham) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStackStruct.scala:167]: That's also made be a little nervous that I've missed something with other implementations. I'll do another pass.

**Review by chrisvittal — APPROVED** [review]

**Pull request merged**